packages/@markuplint/ml-core/src/ml-dom/document.ts

```diff
@@ -78,10 +78,33 @@ export default class MLDOMDocument<T extends RuleConfigValue, O = null> {
 		return treeRoots;
 	}
 
-	async walk(walker: Walker<T, O>) {
-		for (const node of this.nodeList) {
-			await walker(node);
-		}
+	walk(walker: Walker<T, O>): void | Promise<void> {
+		let _resolve: () => void;
+		let _reject: (reason: unknown) => void;
+
+		const promise = new Promise<void>((resolve, reject) => {
+			_resolve = resolve;
+			_reject = reject;
+		});
+
+		const loop = (index = 0) => {
+			if (index >= this.nodeList.length) {
+				_resolve();
+				return;
+			}
+
+			const node = this.nodeList[index];
+			const result = walker(node);
+			if (result instanceof Promise) {
+				result.then(() => loop(index + 1)).catch(_reject);
+			} else {
+				loop(index + 1);
+			}
+		};
+
+		loop();
+
+		return promise;
```

Comment:
@JounQin 

This `return promise` means own function is an async function.
I guess it didn't need change.
Use `await`.

Reply:
@YusukeHirao 

You're right, changed in [`7b24467` (#157)](https://github.com/markuplint/markuplint/pull/157/commits/7b244678bb18da26230fcc91321558712b365390).

Reply:
```js
const testAsync = walker => {
  let _resolve
  let _reject

  const promise = new Promise((resolve, reject) => {
    _resolve = resolve
    _reject = reject
  })

  const nodeList = [1, 2, 3]

  const loop = (index = 0) => {
    if (index >= nodeList.length) {
      return _resolve()
    }
    const node = nodeList[index]
    const result = walker(node)
    if (result instanceof Promise) {
      console.log('async walker:', result)
      result.then(() => loop(index + 1)).catch(_reject)
    } else {
      console.log('sync waker:', result)
      loop(index + 1)
    }
  }

  loop()

  return promise
}

testAsync(node => {
  console.log('node:', node)
  return node
})

testAsync(async node => {
  console.log('node:', node)
  return node
})

console.log('sync:', 1)
```

The above code shows the difference more obviously, you can test it in node fast.

```log
node: 1
sync waker: 1
node: 2
sync waker: 2
node: 3
sync waker: 3
node: 1
async walker: Promise { 1 }
sync: 1
node: 2
async walker: Promise { 2 }
node: 3
async walker: Promise { 3 }
```

Reply:
@JounQin 

Same:

```js
const testAsync = async walker => {
	const nodeList = [1, 2, 3];
	for (const item of nodeList) {
		await walker(item);
	}
	return;
};
```

Reply:
@YusukeHirao 

```js
const sleep = (timeout = 0) =>
  new Promise(resolve => setTimeout(resolve, timeout))

const testAsync = async walker => {
  const nodeList = [1, 2, 3]
  for (const item of nodeList) {
    await walker(item)
  }
}

testAsync(node => {
  console.log('sync node:', node)
  return node
})

testAsync(async node => {
  await sleep(Math.random() * 1000)
  console.log('async node:', node)
  return node
})

console.log('sync:', 0)
```

```log
sync node: 1
sync: 0
sync node: 2
sync node: 3
async node: 1
async node: 2
async node: 3
```

It not correct, `sync: 0` should be logged after `sync node: 3`

Reply:
@JounQin 

It is the same internally.
It must wait for the promise whatever the case finally.

Apparently, I guess it impossible to add the sync APIs until providing sync functions that alternated from all async functions.
And I cannot accept it because my maintenance cost becomes 2 times.

I expected the sync APIs idea that is low cost.
It's disappointing.

Reply:
@YusukeHirao 

> It is the same internally.
> It must wait for the promise whatever the case finally.

No, `verify` of rule is called by `verify` and `verifySync` separately.

When a rule is aync but called by `verifySync`, it just throws at https://github.com/markuplint/markuplint/pull/157/files#diff-258266a7d4660e36f81a70bdad016431829bc10e49a8f129095791ffffd0a2f0R57-R59

> Apparently, I guess it impossible to add the sync APIs until providing sync functions that alternated from all async functions.

I'm not sure to understand, the logic of rules expect `rule-textlint` are all synchronized, although they are marked as `async`, so I removed the `async/await` from rules.

> And I cannot accept it because my maintenance cost becomes 2 times.

How would it cost you to maintain twice? The PR is all the things to be changed, no rule should be changed except removing `async/await`.

Reply:
@JounQin 

I must change sync APIs every if I want to change any inner logic in async APIs. There are 800 lines over.
Umm..., I DO NOT WANT THAT.

Your idea goes to force that.

Reply:
@YusukeHirao 

Yeah, that's right, `async` and `sync` are very different.

I've mentioned that *I'm just doing coping and pasting for now*, maybe the same logic between some of them can be refactored, maybe not. Most of the cope-paste codes are for `file-resolver`. Ideally they should not be changed often.

But of course, if you don't like my current idea, this PR can be closed.

A better approach would be awesome too.

Reply:
Another option is reverting file resolver changes but keep ability for sync rule, and then I'll create an package maybe `markuplint-sync` to extend the ability for linting synchronously.

Reply:
@YusukeHirao 

How do you think? I'd like to work on this approach tonight.

Reply:
@JounQin 

Oh, I see.
This package is under the MIT license.
You create the package is freedom.
I don't have the right to stop you.

If you created it, I want to see it.